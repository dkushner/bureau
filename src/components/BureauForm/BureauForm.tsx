import { Component, Vue, Prop, Provide } from 'vue-property-decorator'
import { VNode, CreateElement } from 'vue'
import { namespace } from 'vuex-class'
import { FormProxy } from './index'

const Bureau = namespace('bureau')

@Component
export default class BureauForm extends Vue {
  @Prop({ type: String, required: true }) name!: string
  @Prop([Array, Function]) validate!: []
  @Prop(Function) onSubmit!: () => void
  @Prop(Boolean) eagerDestroy: boolean = false
  @Prop(Boolean) keepDirty: boolean = false
  @Prop(Object) initialValues!: { [key: string]: any }

  @Bureau.Getter getRegisteredForms!: any
  @Bureau.Getter getForm!: any
  @Bureau.Getter getFormFields!: any
  @Bureau.Getter getFormField!: any
  @Bureau.Getter getFormValues!: any
  @Bureau.Mutation change!: any
  @Bureau.Mutation initialize!: any
  @Bureau.Mutation registerField!: any
  @Bureau.Mutation unregisterField!: any
  @Bureau.Mutation focus!: any
  @Bureau.Mutation blur!: any

  @Provide()
  form: FormProxy = {
    registerField: (fieldName, validators) => {
      this.registerField({ formName: this.name, fieldName })

      if (validators) {
        this.validators[fieldName] = validators
      }
    },
    unregisterField: (fieldName: string) => {
      delete this.validators[fieldName]
      this.unregisterField({ 
        formName: this.name, 
        fieldName, 
        destroy: this.eagerDestroy 
      })
    },
    change: (fieldName: string, value: any) => {
      this.change({ formName: this.name, fieldName, value })
    },
    focus: (fieldName: string) => {
      this.focus({ formName: this.name, fieldName })
    },
    blur: (fieldName: string) => {
      this.blur({ formName: this.name, fieldName })
    },
    getFieldState: (fieldName: string) => {
      return this.getFormField(this.name, fieldName)
    },
    getValue: (fieldName: string) => {
      return this.getFormValues(this.name)[fieldName]
    }
  }

  validators: { [key: string]: any } = { }

  handleSubmit () {
    this.onSubmit()
  }

  created () {
    this.initialize({
      formName: this.name,
      values: this.initialValues,
      keepDirty: this.keepDirty
    })
  }

  render (h: CreateElement): VNode {
    const bindings = {
      on: {
        submit: this.handleSubmit
      }
    }

    return (
      <form { ...bindings }>
        { this.$slots.default }
      </form>
    )
  }
 }