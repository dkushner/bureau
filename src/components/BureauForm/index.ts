import BureauForm from './BureauForm'
import { FieldState } from '../../state';

export interface FormProxy {
  registerField (fieldName: string, validators: any[]): void
  unregisterField (fieldName: string): void
  change (fieldName: string, value: any): void
  focus (fieldName: string): void
  blur (fieldName: string): void
  getFieldState (fieldName: string): FieldState
  getValue (fieldName: string): any
}

export { BureauForm }
export default BureauForm