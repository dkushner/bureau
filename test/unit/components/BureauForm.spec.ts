import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Bureau from '@/index'
import BureauForm from '@/components/BureauForm'
import { mockStore, Overrides } from '../../index'

describe('BureauForm', () => {
  
  test('calls provided submit handler on submit', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const store = mockStore()

    const propsData = {
      name: 'test',
      onSubmit: jest.fn()
    }

    const wrapper = mount(BureauForm, { 
      localVue, 
      store,
      propsData
    })

    const form = wrapper.find('form')
    form.trigger('submit')

    expect(propsData.onSubmit).toBeCalled()
  })

  test('registers form state on creation', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const mutations = {
      initialize: jest.fn()
    }

    const store = mockStore({ mutations })

    const propsData = {
      name: 'test'
    }

    const wrapper = mount(BureauForm, {
      localVue,
      store,
      propsData
    })

    expect(mutations.initialize).toBeCalled()
    expect(store.state).toHaveProperty('bureau.forms')
  })

  test('renders as expected', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const store = mockStore()

    const propsData = {
      name: 'test'
    }

    const wrapper = mount(BureauForm, {
      localVue,
      store,
      propsData
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})