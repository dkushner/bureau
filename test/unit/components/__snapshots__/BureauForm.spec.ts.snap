// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BureauForm renders as expected 1`] = `

<form>
</form>

`;
